components/net/local_directory_listing.rs

```diff
@@ -0,0 +1,405 @@
+        // linked directories and files correctly when clicked.
+        if let Ok(dir_url) = Url::from_directory_path(url.path()) {
+            ServoUrl::from_url(dir_url)
+        } else {
+            return Response::network_error(NetworkError::Internal(format!(
+                "Unable to parse local directory path {}",
+                url
+            )));
+        }
+    } else {
+        url
+    };
+    let mut response = Response::new(url, ResourceFetchTiming::new(request.timing_type()));
+    response.headers.typed_insert(ContentType::html());
+    let (path, has_parent, items) = summarise_directory(path_buf);
+    let page_text = build_html_directory_listing(path, has_parent, items);
+    let bytes: Vec<u8> = page_text.into_bytes();
+    *response.body.lock().unwrap() = ResponseBody::Done(bytes);
+    response
+}
+
+pub fn summarise_directory(
+    value: PathBuf,
+) -> (
+    Result<String, &'static str>,
+    bool,
+    Result<BTreeMap<OsString, DirectoryItem>, &'static str>,
+) {
+    let path = value
+        .to_str()
+        .map(str::to_string)
+        .ok_or("Invalid directory path.");
+    let has_parent = value.parent().is_some();
+    let items = if let Ok(entries) = std::fs::read_dir(value) {
+        Ok(gather_directory_items(entries))
+    } else {
+        Err("Unable to iterate directory contents.")
+    };
+    (path, has_parent, items)
+}
+
+fn gather_directory_items(entries: std::fs::ReadDir) -> BTreeMap<OsString, DirectoryItem> {
+    let map: BTreeMap<OsString, DirectoryItem> = entries
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .flat_map(|e| create_name_item_mapping(e))
+        .collect();
+    map
+}
+
+fn create_name_item_mapping(entry: DirEntry) -> Option<(OsString, DirectoryItem)> {
+    if let Ok(meta) = entry.metadata() {
+        let os_name = entry.file_name();
+        let entry_name = os_name.to_str().map(str::to_string);
+        create_directory_item(entry_name, meta).map(|i| (os_name, i))
+    } else {
+        None
+    }
+}
+
+fn create_directory_item(name: Option<String>, meta: Metadata) -> Option<DirectoryItem> {
+    let last_modified = meta.modified().map(|m| m.into());
+    if meta.is_dir() {
+        Some(DirectoryItem::SubDirectory {
+            name,
+            last_modified,
+        })
+    } else if meta.is_file() || meta.is_symlink() {
+        Some(DirectoryItem::File {
+            is_symlink: meta.is_symlink(),
+            name,
+            size: meta.len(),
+            last_modified,
+        })
+    } else {
+        None
+    }
+}
+
+pub enum DirectoryItem {
+    SubDirectory {
+        name: Option<String>,
+        last_modified: io::Result<OffsetDateTime>,
+    },
+    File {
+        is_symlink: bool,
+        name: Option<String>,
+        size: u64,
+        last_modified: io::Result<OffsetDateTime>,
+    },
+}
+
+// Returns an HTML5 document describing the content of the given local directory.
```

Comment:
triple slash for docs pls. Also, just HTML will do. See https://html.spec.whatwg.org/#is-this-html5?

Reply:
This function is only marked `pub` because I still can't work out how to make it visible to the unit tests without it being visible to the entire project. The intention was to have it non-public, and just give it a basic, short comment.

But, seeing as it is currently visible to all, I'll upgrade the comment to a full rustdoc comment.